packages/react-devtools-extensions/src/parseHookNames/loadSourceAndMetadata.js

```diff
@@ -0,0 +1,432 @@
+//
+// Some overhead may be incurred sharing the source data with the Worker,
+// but less than fetching the file to begin with,
+// and in some cases we can avoid serializing the source code at all
+// (e.g. when we are in an environment that supports our custom metadata format).
+//
+// The overall flow of this file is such:
+// 1. Find the Set of source files defining the hooks and load them all.
+//    Then for each source file, do the following:
+//
+//    a. Search loaded source file to see if a custom React metadata file is available.
+//       If so, load that file and pass it to a Worker for parsing and extracting.
+//       This is the fastest option since our custom metadata file is much smaller than a full source map,
+//       and there is no need to convert runtime code to the original source.
+//
+//    b. If no React metadata, search loaded source file to see if a source map is available.
+//       If so, load that file and pass it to a Worker for parsing.
+//       The source map is used to retrieve the original source,
+//       which is then also parsed in the Worker to infer hook names.
+//       This is less ideal because parsing a full source map is slower,
+//       since we need to evaluate the mappings in order to map the runtime code to the original source,
+//       but at least the eventual source that we parse to an AST is small/fast.
+//
+//    c. If no source map, pass the full source to a Worker for parsing.
+//       Use the source to infer hook names.
+//       This is the least optimal route as parsing the full source is very CPU intensive.
+
+import {__DEBUG__} from 'react-devtools-shared/src/constants';
+import {getHookSourceLocationKey} from 'react-devtools-shared/src/hookNamesCache';
+import {sourceMapIncludesSource} from '../SourceMapUtils';
+import {
+  withAsyncPerformanceMark,
+  withCallbackPerformanceMark,
+  withSyncPerformanceMark,
+} from 'react-devtools-shared/src/PerformanceMarks';
+
+import type {
+  HooksNode,
+  HookSource,
+  HooksTree,
+} from 'react-debug-tools/src/ReactDebugHooks';
+import type {MixedSourceMap} from 'react-devtools-extensions/src/SourceMapTypes';
+
+// Prefer a cached albeit stale response to reduce download time.
+// We wouldn't want to load/parse a newer version of the source (even if one existed).
+const FETCH_WITH_CACHE_OPTIONS = {cache: 'force-cache'};
+
+const MAX_SOURCE_LENGTH = 100_000_000;
+
+export type HookSourceAndMetadata = {|
+  // Generated by react-debug-tools.
+  hookSource: HookSource,
+
+  // Compiled code (React components or custom hooks) containing primitive hook calls.
+  runtimeSourceCode: string | null,
+
+  // Same as hookSource.fileName but guaranteed to be non-null.
+  runtimeSourceURL: string,
+
+  // Raw source map JSON.
+  // Either decoded from an inline source map or loaded from an externa source map file.
+  // Sources without source maps won't have this.
+  sourceMapJSON: MixedSourceMap | null,
```

Comment:
we don't create the SourceConsumer / MetadataConsumer in this file because we'd need to serialize them over to the worker thread, right?

Reply:
Right, and we couldn't do that b'c they're not primitive types.